### ISSUE TITLE
systemd user configuration broken after system update, makes session fail

### ISSUE THREAD

**Issue by xunam** [NixOS/nixpkgs#114140]
**Describe the bug**
When enabling user services with `systemctl --user enable`, symlinks are created in the user's home directory to files in the Nix store. If an upgrade changes the versions of the services, the symlinks may break, making the user instance of systemd unable to start. As a consequence, dbus is not started and various services stop working.

**To Reproduce**
1. Enable a user service, e.g. `systemctl --user enable redshift`
2. Upgrade NixOS to a version with a different derivation for `systemd`
3. Collect garbage with `nix-collect-garbage -d`
4. Log out and try to log back in.

**Expected behavior**
User session starts, possibly with some user services missing.

**Screenshots**
Not applicable.

**Additional context**
My default environment is Sway started from GDM and this bug makes the session unable to start: after entering my password, I am thrown back to the login screen. I can still log in with a Gnome session but anything that requires DBus is non functional.

I don't know exactly which symlink is the problem here. I had the following broken links in `~/.config/systemd/user/`:
- `default.target` points to `/nix/store/XXXX-systemd-246.6/example/systemd/user/default.target`
- `redshift.service` and `default.target.wants/redshift.service` both point to the same `/nix/store/XXXX-unit-redshift.service/redshift.service`
I didn't try to fix only `default.target`.

Obviously these links are not seen by the garbage collector and this should be fixed somehow.

Besides, the fact that it prevents the session from starting suggests a issue with systemd upstream.

**Notify maintainers**

Should I notify @andir @edolstra @flokli as maintainers of the `systemd` package?

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.99, NixOS, 20.09.3246.d4189f68fdb (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(manu): `"nixpkgs-21.05pre271812.58c19a2b0bb"`
 - channels(root): `"nixos-20.09.3246.d4189f68fdb"`
 - nixpkgs: `/home/manu/.nix-defexpr/channels/nixpkgs`

**Comment by flokli** [comment]
If you manually run `systemctl --user enable …`, it'll create a file in your homedir pointing to some location in the nix store, which will eventually get garbage collected. That's somewhat expected. The underlying problem is that there's currently no declarative concept of per-user-enabled units in nixpkgs (only for packages).

Related to https://github.com/NixOS/nixpkgs/issues/108643#issuecomment-756172827.

**Comment by xunam** [comment]
I understand the source of the issue and the fact that this behaviour can be expected by someone with some knowledge of how Nix works. And certainly, having a clear concept of user environment that would include services etc while integrating with the Nix infrastructure, as home-manager tries to achieve as far as I understand, would be nice.

But I feel there is something more here, because using a tool as it is intended to be used leads to breakage. It is part of the design of systemd that a user can `systemctl --user enable` something, and it is to be expected that users will do this. Probably even more so in a multi-user system where the administrator chooses NixOS while users have no particular knowledge of it and just expect a working Linux environment. There are certainly even cases where user services are enabled by other tools than the command-line `systemctl` (GUIs probably do that sometimes). The issue does not occur for system-wide services both because `/etc/systemd` is read-only (so the command will fail) and because the administrator will use `/etc/nixos/configuration.nix` anyway.

I can imagine various ways to address the issue, but I don't know if any one of them is both doable and a good idea:
- make it so that created symlinks will point to `/run/current-system/...` instead of directly to the Nix store (could fix the issue in case of upgrades but not if a package happens to be uninstalled)
- issue a big warning when enabling user services (does not fix anything but the the user will be aware of potential problems)
- disable `systemctl --user enable` completely (feels a bit too radical)
- add an entry in `gcroots` when a link in created, the way `nix-build` does

Any thoughts on these options ?

**Comment by flokli** [comment]
1) This can still break, due to them suddenly not being available in a later generation
2) Maybe, but why not better documenting it?
3) Agreed, to drastical. Also, home-manager etc. are using this to manage enabled units there
4) I /think/ home manager also does that.

Maybe parts of this can be solved by documenting things cleaner, and integrating some of the home-manager bits into nixpkgs?

**Comment by csinko** [comment]
I ran into this issue myself and it took a while to figure out what was going on.  This issue goes beyond garbage collecting...

1. Build and add user service X version 1 to my nix environment which is preset enabled
2. Run `systemctl --user enable <service>` -> Symlink is created in `.config/systemd/user` DIRECTLY to service X version 1 unit derivation
3. Update user service X to version 2-> makes new derivation for user-units that `/etc/static/systemd/user` symlinks to.  This new derivation links to the new user service X version 2
4. When the nixos rebuild switch is executed, systemd reloads, but `.config/systemd/user` overrides `/etc/systemd/user` (which symlinks  to `/etc/static/systemd/user`, so systemd still runs my service X version 1 service, even though I updated to version 2.

The symlinks in `.config/systemd/user` should also be updated if the service has a new derivation IMO.  Otherwise, you are unexpectedly pinned to the old version unless you run `systemctl --user disable <service> && systemctl --user enable <service>` which is confusing.  The garbage collection will also break the system, as stated above, which is all the more reason to force an update here.

**Comment by dennybaa** [comment]
Hello. It's not just nixos - any generic multiuser nix installation is affected.

```
❯ systemctl --user enable podman.socket
Created symlink /home/dennybaa/.config/systemd/user/podman.socket → /nix/store/kdqhmyvzbx5pyc0lkicrjj387j21p9w3-podman-5.7.0/share/systemd/user/podman.socket.
Created symlink /home/dennybaa/.config/systemd/user/sockets.target.wants/podman.socket → /nix/store/kdqhmyvzbx5pyc0lkicrjj387j21p9w3-podman-5.7.0/share/systemd/user/podman.socket.
```

Are there any good mitigations for this?

**Comment by eclairevoyant** [comment]
Yes, the answer is to not use imperative commands to point at changeable symlinks, but rather actually use your config to manage these things. `systemctl enable` is inadvisable on NixOS or any system where you're using service files provided by nix.

**Comment by corngood** [comment]
> Yes, the answer is to not use imperative commands to point at changeable symlinks, but rather actually use your config to manage these things.

I think we should be able to point to how to do this, if it's even possible.

> The underlying problem is that there's currently no declarative concept of per-user-enabled units in nixpkgs (only for packages).

Is this still the case?

For now I've hacked around this by overwriting the links:

```
~/.config/systemd/user/syncthing.service -> /etc/systemd/user/syncthing.service
~/.config/systemd/user/default.target.wants/syncthing.service -> /etc/systemd/user/syncthing.service
```

**Comment by eclairevoyant** [comment]
> I think we should be able to point to how to do this, if it's even possible.

It's not, is the point. Absolute paths will change by generation. That's why nix-based user-management solutions exist.
